ush/python/pygfs/task/land_analysis.py

```diff
@@ -129,26 +146,149 @@ def prepare_IMS(self: Analysis) -> None:
 
         # Ensure the IODA snow depth IMS file is produced by the IODA converter
         # If so, copy to COM_OBS/
-        if not os.path.isfile(f"{os.path.join(self.task_config.DATA, output_file)}"):
+        if not os.path.isfile(f"{os.path.join(localconf.DATA, output_file)}"):
             logger.exception(f"{self.task_config.IMS2IODACONV} failed to produce {output_file}")
-            raise FileNotFoundError(f"{os.path.join(self.task_config.DATA, output_file)}")
+            raise FileNotFoundError(f"{os.path.join(localconf.DATA, output_file)}")
         else:
             logger.info(f"Copy {output_file} to {self.task_config.COM_OBS}")
             FileHandler(prep_ims_config.ims2ioda).sync()
 
     @logit(logger)
-    def get_bkg_dict(self, config: Dict) -> Dict[str, List[str]]:
+    def initialize(self) -> None:
+        """Initialize method for Land analysis
+        This method:
+        - creates artifacts in the DATA directory by copying fix files
+        - creates the JEDI LETKF yaml from the template
+        - stages backgrounds, observations and ensemble members
+
+        Parameters
+        ----------
+        self : Analysis
+            Instance of the LandAnalysis object
+        """
+
+        super().initialize()
```

Comment:
This copies the observation data in the IODA format based the predefined observation yaml file. 
However, the IMS observation data in the final IODA format (`gdas.t{HH}z.ims_snow_{PDY}{HH}.nc4`) is not available yet, and will be computed in `prepare_IMS`. The name convention of the original IMS data is `ims{YYYY}{DOY}_4km_v1.3.nc`. Therefore, the following errors were caused: 

```
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/ush/python/pygw/src/pygw/fsutils.py", line 83, in cp
    shutil.copyfile(source, target)
  File "/scratch1/NCEPDEV/da/python/opt/core/miniconda3/4.6.14/envs/gdasapp/lib/python3.7/shutil.py", line 120, in copyfile
    with open(src, 'rb') as fsrc:
FileNotFoundError: [Errno 2] No such file or directory: '/scratch1/NCEPDEV/global/Jiarui.Dong/ptmp/landid8888/gdas.20210323/18/obs/gdas.t18z.ims_snow_2021032318.nc4'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/scripts/exglobal_land_analysis.py", line 23, in <module>
    anl.initialize()
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/ush/python/pygw/src/pygw/logger.py", line 264, in wrapper
    retval = func(*args, **kwargs)
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/ush/python/pygfs/task/land_analysis.py", line 170, in initialize
    super().initialize()
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/ush/python/pygfs/task/analysis.py", line 37, in initialize
    FileHandler(obs_dict).sync()
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/ush/python/pygw/src/pygw/file_utils.py", line 39, in sync
    sync_factory[action](files)
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/ush/python/pygw/src/pygw/file_utils.py", line 59, in _copy_files
    cp(src, dest)
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/ush/python/pygw/src/pygw/fsutils.py", line 85, in cp
    raise OSError(f"unable to copy {source} to {target}")
OSError: unable to copy /scratch1/NCEPDEV/global/Jiarui.Dong/ptmp/landid8888/gdas.20210323/18/obs/gdas.t18z.ims_snow_2021032318.nc4 to /scratch1/NCEPDEV/stmp2/Jiarui.Dong/RUNDIRS/landid8888/landanl.135218/obs/gdas.t18z.ims_snow_2021032318.nc4
```
One solution is to rename the original IMS data name from `ims{YYYY}{DOY}_4km_v1.3.nc` to `gdas.t{HH}z.ims_snow_{PDY}{HH}.nc4`. Any other solutions? Thanks.

Reply:
I can't reproduce this error.  Can we discuss offline what the issue is?  
I thought the observations are provided in a yaml.

Reply:
In ush/python/pygfs/task/analysis.py:

```
def get_obs_dict(self) -> Dict[str, Any]:

        # get observers from master dictionary
        observers = obs_list_config['observers']
        copylist = []
        for ob in observers:
            obfile = ob['obs space']['obsdatain']['engine']['obsfile']
            basename = os.path.basename(obfile)
            copylist.append([os.path.join(self.task_config['COM_OBS'], basename), obfile])
        obs_dict = {
            'mkdir': [os.path.join(self.runtime_config['DATA'], 'obs')],
            'copy': copylist
        }
        return obs_dict
```

Reply:
> The observation is provided at the yaml file sorc/gdas.cd/parm/land/obs/config/ims_snow.yaml:
> 
> ```
>   obsdatain:
>     engine:
>       type: H5File
>       obsfile: $(DATA)/obs/$(OPREFIX)ims_snow_{{ current_cycle | to_YMDH }}.nc4
> ```
> 
> In the t00z, t06z, and t12z cycles, `ims_snow.yaml` is not called. Only in the t18z cycle, the ims snow data is required.

It seems that the observation file needs to be cycle dependent.

Reply:
Or you can link the `ims{YYYY}{DOY}_4km_v1.3.nc` as `gdas.t{HH}z.ims_snow_{PDY}{HH}.nc4` for all cycles except `18z`.
This will allow you to retain the same observation yaml.

Reply:
We already made the observation yaml file as cycle dependent in config.landanl as: 

```
obs_list_name=gdas_land_adpsfc_only.yaml
if [[ "${cyc}" = "18" ]]; then
    obs_list_name=gdas_land_prototype.yaml
fi
```

While:
**gdas_land_adpsfc_only.yaml:**
```
observers:
- !INC ${OBS_YAML_DIR}/adpsfc_snow.yaml
```

**gdas_land_prototype.yaml:**
```
observers:
- !INC ${OBS_YAML_DIR}/adpsfc_snow.yaml
- !INC ${OBS_YAML_DIR}/ims_snow.yaml
```

Therefore, there are no any issues for the t00z, t06z and t12z cycles. The problem is only at the t18z cycle when `ims_snow.yaml` is included. As you suggested, I can link the `ims{YYYY}{DOY}_4km_v1.3.nc` as `gdas.t{HH}z.ims_snow_{PDY}{HH}.nc4` to avoid this problem, and later the `gdas.t{HH}z.ims_snow_{PDY}{HH}.nc4` will be replaced by the computed file from prepare_IMS. @CoryMartin-NOAA Do you think it is okay for this approach? Thanks.

Reply:
What we really need is IODA to handle missing obs... I would rather not link a wrong file here because that is asking for trouble. We can create a fake file if you want, but I think that is more workflow effort than just doing what @jiaruidong2017 mentioned above and use a different list YAML depending on cycle.

Reply:
I suspect simply linking `ims{YYYY}{DOY}_4km_v1.3.nc` as `gdas.t{HH}z.ims_snow_{PDY}{HH}.nc4` won't be enough.  IODA is not just looking for the file to be present, but it also expects it to be a IODA file.  `ims{YYYY}{DOY}_4km_v1.3.nc` is not a IODA file and so the program will abort. 

We can proceed with the approach of including the `ims_snow.yaml` only for the 18z cycle.

@CoryMartin-NOAA Is there an issue open in IODA that communicates this requirement for IODA to handle missing IODA files?  If not, I would recommend opening an issue.

Reply:
Yes https://github.com/JCSDA-internal/ioda/issues/541

Reply:
This issue has been fixed with the change at https://github.com/NOAA-EMC/GDASApp/pull/490#issuecomment-1563363699